src/Microsoft.AspNetCore.Blazor/Rendering/Renderer.cs

```diff
@@ -121,25 +123,53 @@ internal void AssignEventHandlerId(ref RenderTreeFrame frame)
             // In order to dispatch the event, we need a UIEventHandler, so we're going weakly
             // typed here. The user will get a cast exception if they map the wrong type of
             // delegate to the event.
-            if (frame.AttributeValue is UIEventHandler wrapper)
+            if (frame.AttributeValue is Func<UIEventArgs, Task> handler)
             {
-                _eventHandlersById.Add(id, wrapper);
+                _eventHandlersById.Add(id, handler);
             }
+
             // IMPORTANT: we're creating an additional delegate when necessary. This is
             // going to get cached in _eventHandlersById, but the render tree diff
             // will operate on 'AttributeValue' which means that we'll only create a new
             // wrapper delegate when the underlying delegate changes.
             //
             // TLDR: If the component uses a method group or a non-capturing lambda
             // we don't allocate much.
+            else if (frame.AttributeValue is UIEventHandler wrapper)
+            {
+                _eventHandlersById.Add(id, (e) => 
+                {
+                    wrapper(e);
+                    return Task.CompletedTask;
+                });
+            }
             else if (frame.AttributeValue is Action action)
             {
-                _eventHandlersById.Add(id, (UIEventArgs e) => action());
+                _eventHandlersById.Add(id, (e) =>
+                {
+                    action();
+                    return Task.CompletedTask;
+                });
+            }
+            else if (frame.AttributeValue is Func<Task> asyncAction)
+            {
+                _eventHandlersById.Add(id, (e) =>
+                {
+                    return asyncAction();
+                });
             }
             else if (frame.AttributeValue is MulticastDelegate @delegate)
             {
-
-               _eventHandlersById.Add(id, (UIEventArgs e) => @delegate.DynamicInvoke(e));
+               _eventHandlersById.Add(id, (UIEventArgs e) => 
+               {
+                   var value = @delegate.DynamicInvoke(e);
+                   if (value is Task task)
+                   {
+                       return task;
+                   }
+
+                   return null;
+               });
```

Comment:
One idea that I had about this ugliness, there are basically 4 types of signatures we want to support. We can abstract this behind a struct `EventBinding` so that we have an exchange type we can pass to `DispatchEvent`. Then we don't need to allocate these wrapper delegates to make everything callable via a `Func<UIEventArgs, Task>` without any allocations - since the struct will do that for us.

Reply:
I don't think we'll be able to get rid of this cast ladder totally. I'll wait to speculate more until I try it out.